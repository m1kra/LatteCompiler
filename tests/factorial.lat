int iter_factorial(int n){
    int i = 1, res = 1;
    while (i < n + 1){
        res = res * i;
        i++;
    }
    return res;
}

int rec_factorial(int n){
    if (n < 2)
        return 1;
    return n * rec_factorial(n - 1);
}

int rec2a_factorial(int n){
    if (n < 2)
        return 1;
    return n * rec2b_factorial(n - 1);
}

int rec2b_factorial(int n){
    if (n < 2)
        return 1;
    return n * rec2a_factorial(n - 1);
}


void main(){
    printInt(iter_factorial(5));
    printInt(rec_factorial(5));
    printInt(rec2a_factorial(5));
    printInt(rec_factorial(12));
    printInt(iter_factorial(rec_factorial(3)));
}