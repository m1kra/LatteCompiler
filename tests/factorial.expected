120
120
120
479001600
720